odoo/addons/base/ir/ir_mail_server.py

```diff
@@ -464,6 +465,70 @@ def send_email(self, message, mail_server_id=None, smtp_server=None, smtp_port=N
             raise MailDeliveryException(_("Mail Delivery Failed"), msg)
         return message_id
 
+    @api.multi
+    def _get_rewritten_from(self, original_from):
+        """Returns a FROM header to respect the SMTP server settings.
+
+        Args:
+            original_from (str): The original FROM header for the message.
+
+        Returns:
+            str: A string that should be used in a message header.
+        """
+        # Use the canonical name, if existing.
+        address = parseaddr(original_from)
+        # Otherwise we use the name part of the email address.
+        if not address[0]:
+            address[0] = address[1].split('@', 1)[0]
+        rewritten_name = _('%(author)s via %(company)s') % {
+            'author': address[0],
+            'company': self.env.user.company_id.name,
+        }
+        return formataddr((rewritten_name, self._get_default_bounce_address()))
+
+    @api.model
+    def _get_rewrite_from_whitelist(self):
+        """Return a list of domains that this system is allowed to send from.
+
+        Returns:
+            list(str): Domains that this system can send authenticated emails
+                for.
+        """
+        whitelist = self.env['ir.config_parameter'].sudo().get_param(
+            'mail.rewrite_from.whitelist', default='',
+        )
+        if whitelist.strip():
+            return [line.strip() for line in whitelist.split(',')]
+
+    @api.multi
+    def _update_message_from(self, message):
+        """Update the FROM header on message to align with server settings.
+
+        Args:
+            message (email.message): The message object to be evaluated and
+                possibly mutated.
+        """
+
+        self.ensure_one()
+
+        domain_catchall = self.env['ir.config_parameter'].sudo().get_param(
+            'mail.catchall.domain',
+        )
+        # Without a catchall domain, we do not have an address to send from
+        if not domain_catchall:
+            return
+
+        whitelist = self._get_rewrite_from_whitelist()
+        address = email_split(message['From'])[0]
+        domain_email = address.split('@', 1)[1]
+
+        if not whitelist or domain_email in whitelist:
```

Comment:
I think this condition is not the right one. Something like:
```python
if whitelist and domain_email in whitelist:
    return
```
should be better. Because otherwise, we do want to rewrite the From header...